### ISSUE TITLE
Convert all component command-line flags to versioned configuration

### ISSUE THREAD

**Issue by bgrant0607** [GoogleCloudPlatform/kubernetes#12245]
Forked from #1627, #5472, #246, #12201, and other issues.

We configure Kubernetes components like apiserver and controller-manager with command-line flags currently. Every time we change these flags, we break everyone's turnup automation, at least for anything not in our repo: hosted products, other repos, blog posts, ...

OTOH, the scheduler and kubectl read configuration files. We should be treating all configuration similarly, as versioned APIs, with backward compatibility, deprecation policies, documentation, etc. How we distribute the configuration is a separate issue -- #1627.

cc @davidopp @thockin @dchen1107 @lavalamp

**Comment by bgrant0607** [comment]
Scheduler config, as an example: https://github.com/kubernetes/kubernetes/tree/master/plugin/pkg/scheduler/api

**Comment by bgrant0607** [comment]
And Openshift:
https://github.com/openshift/origin/blob/master/pkg/cmd/server/api/v1/types.go

**Comment by bgrant0607** [comment]
And kubeconfig: https://github.com/kubernetes/kubernetes/tree/master/pkg/client/clientcmd/api

**Comment by stefwalter** [comment]
@mikedanese I started yesterday to convert KubeletServer (which is poorly named) to use a JSON settings file. I'd be happy to follow the lead of your work done on the scheduler config, as it does cover a couple things I missed.

**Comment by mikedanese** [comment]
I haven't done much with this yet. I've begun to add a component config api group in this commit:

https://github.com/kubernetes/kubernetes/commit/7f76a8ed1bb974f2c0ab53d08940ba3bb9cfa2ec

Is this a similar direction you are taking? Let's discuss how we want to redistribute this work tomorrow.

**Comment by pmorie** [comment]
@mikedanese @stefwalter Can you guys confirm that you intend to combine this with the mechanism described in #6477 ?

**Comment by mikedanese** [comment]
Yes it is. It is related but separate and can be done in parallel. The combination should close #1627.

**Comment by mikedanese** [comment]
@stefwalter can you please point me to the branch where you are working on this?

**Comment by bgrant0607** [comment]
All new configuration formats we create should go into new API groups #12951

**Comment by stefwalter** [comment]
@mikedanese I stopped working on it once I saw your work. I did about half an hour ... before posting what I was doing ... https://github.com/stefwalter/kubernetes/tree/kubelet-settings

I'd be happy to help out with the tedious bits or wherever necessary.

**Comment by bgrant0607** [comment]
Some additional motivation:

The current way in which we configure components via command-line flags is unnecessarily convoluted and hard to standardize, since some deployments run the components using containers, some use init, some use systemd, some use supervisord, etc.

Life of configuring a flag on GCE (if wrong, that's another indication of how convoluted it is; different for every cloud platform):

1. Set an environment variable (e.g., ADMISSION_CONTROL) in a config-default.sh script that is invoked by some script invoked indirectly by kube-up.sh: https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/config-default.sh
2. Which is semi-manually translated to a YAML file and stuffed into VM metadata: https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/debian/helper.sh
3. Which is used to generate a Salt pillar on the master node: https://github.com/kubernetes/kubernetes/blob/master/cluster/gce/configure-vm.sh
4. Which is used to template-substitute the apiserver's pod config: https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/kube-apiserver/kube-apiserver.manifest

**Comment by davidopp** [comment]
Not for 1.1

**Comment by davidopp** [comment]
Dropping to P2 - @mikedanese said he thought he'd be able to get everything except API server into v1.2, and I think that's fine.

**Comment by bogdando** [comment]
> configure Kubernetes components like apiserver and controller-manager with command-line flags ...
> is unnecessarily convoluted and hard to standardize, since some deployments run the components using containers, some use init, some use systemd, some use supervisord...

I strongly believe that truly immutable containers shall not only ship immutable binaries and dependencies bundled but as well behave always the same way, which is independent of the mounted
configuration files. I may be wrong and opinionated, but that's how I "feel" containers. And runtime flags do perfectly fit into that concept, unlike to config files.

With config files mounted RW into *immutable* containers, you always end up with risks to make those containers to behave unpredictable, which has nothing to real immutability. When deployed or *mutated* with bad configs, it becomes as well destructive if had been shipped with bad binaries or dependencies. An example failure mode to illustrate that:
- By a mistake in the systemd unit file of a containerized kubelet service, let it be badly written teardown/init for post-stop/pre-start, create two running instances of the kubelet service with mounted config files in RW mode.
- Both of the kublet service instances do checkpointing of its current state (see [self-hosted kuberenetes's](https://github.com/kubernetes/community/blob/master/contributors/design-proposals/self-hosted-kubelet.md) "checkpoint the updated config"
- The RW mounted checkpointed state ends up in the corrupted file accessed by two instances of kubelet
- ???
(nothing good, but pure destruction is now delayed for the next kubelet service restart in order to take over the control plane to the self-hosted one).

Also, mutated config files must notify the related service on changes, this brings in extra moving parts, like classic config management tools and its agents etc. More things to fail.

Given that said, I'd like the community to reconsider abandoning runtime flags, which make the given failure mode impossible and does not require notifications on the config changes, all options are self contained in the service unit definition - just restart it *to atomically mutate and apply*. This brings none moving parts and none extra failure modes.

@mikedanese @bgrant0607

**Comment by bogdando** [comment]
And if one wants versioning, just please do versioning of the self-contained service units files containing all required runtime flags corresponding to the given version.

**Comment by bgrant0607** [comment]
@bogdando Systemd unit files are configuration files in a less introspectable, less portable, less maintainable format. I don't think your argument is actually about flags vs configuration files at all. Read-only vs. read-write seems to be a fairly narrow/specific issue that could be easily addressed. Since this appears to be specifically about Kubelet, please raise your concerns on #29459 or #27980. As for checkpointing in Kubelet, that's inevitable due to availability and bootstrapping requirements. If you have specific suggestions for how to make checkpointing more robust and resilient, please comment on #489.

**Comment by bogdando** [comment]
@bgrant0607 thanks for the links, I will take a look. Note, the runtime flags vs config files bring less corner cases to handle when doing checlpointing, that was my main concern. Also, systemd unit files looks like portable well enough as supported by all modern OS types, don't they?

**Comment by kfox1111** [comment]
I'm not sure if this is relevant or not to the conversation, but if daemonsets are enhanced to support the linked feature, along with daemonset rolling upgrade support, maybe this will be useful:
https://github.com/kubernetes/kubernetes/issues/43112

**Comment by roberthbailey** [comment]
@mikedanese is going to be driving the component configuration standardization for 1.7.

**Comment by mtaufen** [comment]
Guide on how to implement componentconfig: https://docs.google.com/document/d/1FdaEJUEh091qf5B98HM6_8MS764iXrxxigNIdwHYW9c/

**Comment by kfox1111** [comment]
doc is private

**Comment by mtaufen** [comment]
It is shared with the kubernetes-dev group.

**Comment by neolit123** [comment]
this topic has multiple stakeholding sigs, instead of cluster-lifecycle as the owner.
tagging sig-arch as the umbrella.

/sig architecture
/remove-sig cluster-lifecycle

the design idea is already in place and the component owners are graduating their config to v1.

- kube-schduler - v1
- kube-controller-manager - v1alpha1
- kube-apiserver - none yet
- kube-proxy - v1alpha1
- kubeadm - v1beta2
- kubelet - v1beta1

**Comment by neolit123** [comment]
> Apologize for the bothering, I'm wondering if there's any update on the progress of this issue? Recently I found [controller-runtime deprecating its ComponentConfig implementation](https://github.com/kubernetes-sigs/controller-runtime/issues/895), so I comed up with this issue in k/k, and wanted to know the community's current opinion and roadmap about this.

similarly to controller runtime, k/k componentconfig apis have made a rather slow progress into graduation to v1. some api versions are stale and should be higher - i.e. maturity is not accurately represented.

there has not been a discussion to drop the effort entirety.
the apis are widely used, despite not beeing v1 - e.g. kubelet is still at v1beta1. removing such apis will be very disruptive to users.

**Comment by zqzten** [comment]
@neolit123 Thanks for the rapid reply! I would also be happy to see these configurations get maintained, is there anybody or group currently responsible for this?

Also, I've found that not every component of k/k has supported being configured by a versioned configration file, and for kubelet, it implemented the whole processing logic in its own code base which cannot be easily reused. So I'm wondering if there's any chance that we build a common mechanism (e.g. in the `component-base`) which can be reused across all the components of k/k, and even for other sub projects (such as the controller-runtime above). This way we can achieve a unified behavior as well as reducing the overall maintenance cost.

**Comment by neolit123** [comment]
> @neolit123 Thanks for the rapid reply! I would also be happy to see these configurations get maintained, is there anybody or group currently responsible for this?
> 

currently, different groups / sigs own components and their respective configs.

> Also, I've found that not every component of k/k has supported being configured by a versioned configration file, and for kubelet, it implemented the whole processing logic in its own code base which cannot be easily reused. So I'm wondering if there's any chance that we build a common mechanism (e.g. in the `component-base`) which can be reused across all the components of k/k, and even for other sub projects (such as the controller-runtime above). This way we can achieve a unified behavior as well as reducing the overall maintenance cost.

a working group called "component standard" would have been the appropriate forum for this discussion but it disbanded. currently, each component has the freedom to decide on the versioned config implementation details. if you want to open the discussion the next suitable place would be the sig architecture regular zoom call.

https://github.com/kubernetes/community/blob/master/sig-architecture/README.md#meetings

please add the topic to the agenda doc there if you'd like.
